library/alloc/src/boxed.rs

```diff
@@ -1955,11 +1955,37 @@ impl<T: ?Sized + Unsize<U>, U: ?Sized, A: Allocator> CoerceUnsized<Box<U, A>> fo
 #[unstable(feature = "dispatch_from_dyn", issue = "none")]
 impl<T: ?Sized + Unsize<U>, U: ?Sized> DispatchFromDyn<Box<U>> for Box<T, Global> {}
 
+#[cfg(not(no_global_oom_handling))]
+trait SpecFromIter<T, I> {
+    fn from_iter(iter: I) -> Self;
+}
+
+#[cfg(not(no_global_oom_handling))]
+impl<T, I> SpecFromIter<T, I> for Box<[T]>
+where
+    I: Iterator<Item = T>,
+{
+    default fn from_iter(iter: I) -> Self {
+        iter.collect::<Vec<_>>().into_boxed_slice()
+    }
+}
+
+#[cfg(not(no_global_oom_handling))]
+impl<T, I> SpecFromIter<T, I> for Box<[T]>
+where
+    I: Iterator<Item = T> + TrustedLen,
+{
+    fn from_iter(iter: I) -> Self {
+        // SAFETY: `.collect()`ing `TrustedLen` iterator into a `Vec<_>` never reallocates
+        unsafe { iter.collect::<Vec<_>>().into_exact_size_boxed_slice() }
```

Comment:
The specialization for `TrustedLen` in Vec uses `with_capacity`, which does not promise exact allocation. Which means it's not guaranteed that the Vec can be turned into a Box without reallocation.
Either the specialization in `Vec` needs to be changed or (probably the better option) you need to implement a custom collect loop over an uninitialized box.

Reply:
I think it promises exact allocation:

https://doc.rust-lang.org/stable/alloc/vec/struct.Vec.html#method.with_capacity

> The vector will be able to hold _exactly_ capacity elements without reallocating. If capacity is 0, the vector will not allocate.

(emphasis mine)

Reply:
https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#method.with_capacity

> The vector will be able to hold at least capacity elements without reallocating. This method is allowed to allocate for more elements than capacity. If capacity is 0, the vector will not allocate.

Reply:
So documentation changed from `stable` to `nightly`. It’s interesting since this is a breaking change: exact capacity was previously guaranteed, but is not anymore. Furthermore, it contradicts another place in the `Vec<_>` documentation:

> `vec![x; n]`, `vec![a, b, c, d]`, and `Vec::with_capacity(n)`, will all produce a Vec with exactly the requested capacity.

It seems like this change was introduced by @jmaargh in 95dc353, maybe they’ll be able to provide further comment?

Reply:
That may have been an unintended documentation change, I'll ask for clarification.

Reply:
The docs guaranteed exact capacity, but I am fairly certain that was in error (see #95614 and #43490). If the docs are inconsistent that was because I missed something in #96173.

The most obvious case for the error was for ZSTs, but AFAIK the underlying allocator was always allowed to over-allocate resulting in more capacity.

Reply:
I hope this is a case where my pedantic docs change is helping to avoid an error

Reply:
Setting aside ZSTs (they do not seem to matter here nor for any common unsafe usage), this guarantee might be relied upon by unsafe code, so I believe it’s a breaking change to remove it.

While allocator _can_ allocate more than required, this doesn’t matter, since `Vec::with_capacity()` currently always sets requested capacity, ignoring the size of the allocated buffer:

https://github.com/rust-lang/rust/blob/83addf2540ce7ce7d0f875fdbe2a7cd0e7f52d97/library/alloc/src/raw_vec.rs#L198

This might be a bug in `Vec<_>`, but I think breaking changes in the standard library guarantees require some formal process.

Reply:
Side note: `Allocator::shrink()`, which is used by `.shrink_to_fit()`, also doesn’t have a requirement to return an allocation of exact requested size, so if this is unsound, then `.into_boxed_slice()` is _also_ unsound (and probably can’t be implemented at all).

Reply:
I'm not adjudicating on your use of unsafe here,just explaining that the provenance and why of the docs change (that the previously documented guarantee was considered erroneous).

If you think the docs change is wrong, please open a separate issue where it can be discussed.

Reply:
Thanks for clarification. I created a new issue to discuss this.

I believe code in this PR has the same “standard library privileges” as code in `.shrink_to_fit()` and therefore can rely on discarding extra capacity in `RawVec`.

Reply:
Well, `shrink_to_fit` is a different codepath compared to `with_capacity` so they could have different behavior.

Reply:
As far as I can remember, this discussion finished with the decision that while `::with_capacity()` doesn’t promise returning exact capacity (anymore), it promises that it would be possible to deallocate the vector using the capacity that was passed.

Side note: Vec’s documentation still promises that

> `vec![x; n]`, `vec![a, b, c, d]`, and `Vec::with_capacity(n)`, will all produce a `Vec` with exactly the requested capacity. If `len == capacity`, (as is the case for the `vec!` macro), then a `Vec<T>` can be converted to and from a `Box<[T]>` without reallocating or moving the elements.

which I believe to be a mistake in documentation at this point.